agent/eventhandler/attachment_handler.go

```diff
@@ -0,0 +1,158 @@
+	// attachmentARN is the arn of the attachment that the attachmentHandler is handling
+	attachmentARN string
+
+	// stateSaver is a statemanager which may be used to save any
+	// changes to an attachment's SentStatus
+	stateSaver statemanager.Saver
+
+	// backoff is the backoff object used in submitting attachment state change
+	backoff retry.Backoff
+
+	// lock is used to ensure that the attached status of an attachment won't be sent multiple times
+	lock sync.Mutex
+
+	client api.ECSClient
+	ctx    context.Context
+}
+
+// NewAttachmentEventHandler returns a new AttachmentEventHandler object
+func NewAttachmentEventHandler(ctx context.Context,
+	stateSaver statemanager.Saver,
+	client api.ECSClient) *AttachmentEventHandler {
+	return &AttachmentEventHandler{
+		ctx:                    ctx,
+		stateSaver:             stateSaver,
+		client:                 client,
+		attachmentARNToHandler: make(map[string]*attachmentHandler),
+		backoff: retry.NewExponentialBackoff(submitStateBackoffMin, submitStateBackoffMax,
+			submitStateBackoffJitterMultiple, submitStateBackoffMultiple),
+	}
+}
+
+// AddStateChangeEvent adds a state change event to AttachmentEventHandler for it to handle
+func (handler *AttachmentEventHandler) AddStateChangeEvent(change statechange.Event) error {
+	if change.GetEventType() != statechange.AttachmentEvent {
+		return fmt.Errorf("eventhandler: attachment handler received unrecognized event type: %d", change.GetEventType())
+	}
+	event, ok := change.(api.AttachmentStateChange)
+	if !ok {
+		return errors.New("eventhandler: unable to get attachment event from state change event")
+	}
+
+	if event.Attachment == nil {
+		return fmt.Errorf("eventhandler: received malformed attachment state change event: %v", event)
+	}
+
+	attachmentARN := event.Attachment.AttachmentARN
+	handler.lock.Lock()
+	if _, ok := handler.attachmentARNToHandler[attachmentARN]; !ok {
+		handler.attachmentARNToHandler[attachmentARN] = &attachmentHandler{
+			attachmentARN: attachmentARN,
+			stateSaver:    handler.stateSaver,
+			client:        handler.client,
+			ctx:           handler.ctx,
+			backoff:       handler.backoff,
+		}
+	}
+	handler.lock.Unlock()
+
+	attachmentHandler := handler.attachmentARNToHandler[attachmentARN]
+	go attachmentHandler.submitAttachmentEvent(&event)
+
+	return nil
+}
+
+// submitAttachmentEvent submits an attachment event to backend
+func (handler *attachmentHandler) submitAttachmentEvent(attachmentChange *api.AttachmentStateChange) {
+	seelog.Debugf("AttachmentHandler: acquiring attachment lock before sending attachment state change for attachment %s", handler.attachmentARN)
+	handler.lock.Lock()
+	seelog.Debugf("AttachmentHandler: acquired attachment lock for attachment %s", handler.attachmentARN)
+	defer handler.lock.Unlock()
+
+	// we need to continue retrying but skip submitting attachment state change when we successfully submitted the state
+	// change but failed to save state
+	skipSubmitting := false
```

Comment:
Just to clarify, `skipSubmitting` will be used to retry the `state.Save()` operation if there's a failure to save state the first time, yeah? If that is true, I think we should handle it in a different way. If there's an error saving state, more often than not, it means that we've run out of space on the disk. I can't think of many possibilities where saving state fails for the first time, but succeeds the second time. 

With that in mind, a better way to handle this might be retry it may be 3 times and then give up so that we don't hold `handler.lock` for too long here. 

A related question here is: should we attempt to save the state while holding this lock? Can that happen after the fact, without holding the lock?

Reply:
> Just to clarify, `skipSubmitting` will be used to retry the `state.Save()` operation if there's a failure to save state the first time, yeah?
Yes that's what i'm intended to do.

> A related question here is: should we attempt to save the state while holding this lock? Can that happen after the fact, without holding the lock?
I don't think we need to hold the lock when attempting to save state here. Updated to do this in a separate go routine and only retry 3 times for it